packages/grafbase-sdk/README.md

```diff
@@ -0,0 +1,300 @@
+  parent: g.relation(() => user).optional()
+})
+
+// finally we export the default config
+export default config({
+  schema: g
+})
+```
+
+When `grafbase dev` finds the above config from `$PROJECT/grafbase/grafbase.config.ts`, it genereates the SDL to `$PROJECT/.grafbase/generated/schema/schema.graphql`:
+
+```graphql
+type Profile {
+  address: String!
+}
+
+type User @model {
+  name: String!
+  age: Int
+  profile: Profile
+  parent: User
+}
+```
+
+The above SDL is now used when starting the dev.
+
+## Types
+
+Types are generated with the `type` method:
+
+```typescript
+g.type("Profile", {
+  address: g.string()
+})
+```
+
+## Models
+
+Types are generated with the `model` method:
+
+```typescript
+g.model("User", {
+  name: g.string()
+})
+```
+
+## Enums
+
+Enums can be generated either from TypeScript enums or from a dynamic array:
+
+```typescript
+enum Fruits {
+  Apples,
+  Oranges
+}
+
+g.enumType('Fruits', Fruits)
+```
+
+or
+
+```typescript
+g.enumType('Fruits', ['Apples', 'Oranges'])
+```
+
+## Queries and Mutations
+
+Queries are generated with the `query` method, mutations with the `mutation` method:
+
+```typescript
+g.query('greet', {
+  args: { name: g.string().optional() },
+  returns: g.string(),
+  resolver: 'hello'
+})
+
+const input = g.type('CheckoutSessionInput', { name: g.string() })
+const output = g.type('CheckoutSessionOutput', { successful: g.boolean() })
+
+g.mutation('checkout', {
+  args: { input: g.ref(input) },
+  returns: g.ref(output),
+  resolver: 'checkout'
+})
+```
+
+## Unions
+
+Unions can be done using the `union` method:
+
+```typescript
+const user = g.type('User', {
+  name: g.string(),
+  age: g.int().optional()
+})
+
+const address = g.type('Address', {
+  street: g.string().optional()
+})
+
+g.union('UserOrAddress', { user, address })
+```
+
+## Interfaces
+
+Interfaces can be generated using the `interface` method, and a type can be extended with an interface:
+
+```typescript
+const produce = g.interface('Produce', {
+  name: g.string(),
+  quantity: g.int(),
+  price: g.float(),
+  nutrients: g.string().optional().list().optional()
+})
+
+g.type('Fruit', {
+  isSeedless: g.boolean().optional(),
+  ripenessIndicators: g.string().optional().list().optional()
+}).implements(produce)
+```
+
+Notice how one doesn't need to type the fields to the type: they are inferred to the final SDL from the interface definition.
+
+## Field generation
+
+Fields are generated from the `g` object:
+
+- String: `g.string()`
+- ID: `g.id()`
+- Email: `g.email()`
+- Int: `g.int()`
+- Float: `g.float()`
+- Boolean: `g.boolean()`
+- Date: `g.date()`
+- DateTime: `g.datetime()`
+- IPAddress: `g.ipAddress()`
+- Timestamp: `g.timestamp()`
+- URL: `g.url()`
+- JSON: `g.json()`
+
+## Enum fields
+
+```typescript
+// first greate an enum
+enum Fruits {
+  Apples,
+  Oranges
+}
+
+const fruits = g.enumType("Fruits", Fruits)
+
+// then use it e.g. in a model
+g.model("User", {
+  favoriteFruit: g.enum(fruits)
```

Comment:
Could this use `g.ref`, like when you're embedding objects?  Would free up `enum` to be used for the type.